File: solutions/dynamic-programming/322-coin-change.py
"""
# Difficulty: Medium

You are given an integer array coins representing coins of different denominations
and an integer amount representing a total amount of money.

Return the fewest number of coins that you need to make up that amount. If that
amount of money cannot be made up by any combination of the coins, `return -1`.

You may assume that you have an infinite number of each kind of coin.

**Example:**

<dl class="example-details">
<dt>Input:</dt>
<dd>coins = [1,2,5], `amount = 11`</dd>
<dt>Output:</dt>
<dd>3</dd>
<dt>Explanation:</dt>
<dd>Minimum coins to make amount 11 using [1,2,5] is 3 coins: 5+5+1</dd>
</dl>

<details>
<summary><b>🔍 SOLUTION EXPLANATION</b></summary>


### METADATA:
**Techniques**: TBD
**Data Structures**: TBD
**Patterns**: TBD
**Time Complexity**: O(amount × len(coins))
**Space Complexity**: O(amount)

### INTUITION:
This is a classic "minimum path to `target`" DP problem. For any amount, we want
the minimum coins needed. We can build this up from smaller amounts.

### APPROACH:
1. **Initialize DP array**: Create array of size (amount + 1) filled with infinity, representing minimum coins needed for each amount
2. **Set base case**: Set dp[0] = 0 (zero coins needed to make amount 0)
3. **Iterate through amounts**: For each amount from 1 to target, calculate minimum coins needed
4. **Try each coin**: For current amount i, try using each coin denomination that doesn't exceed i
5. **Update DP value**: For each valid coin, calculate dp[i - coin] + 1 and take minimum across all coins
6. **Build up solution**: Each dp[i] is built from previously computed smaller amounts (bottom-up dynamic programming)
7. **Return result**: Return dp[amount] if reachable (not infinity), otherwise return -1 (impossible to make amount)

### WHY THIS WORKS:
Using BFS with a queue processes nodes level by level. Tracking level size ensures we group nodes correctly. This works because BFS naturally visits nodes in level order, and we can identify level boundaries by counting nodes in the queue at each level's start.

### EXAMPLE WALKTHROUGH:
```
coins = [1,2,5], `amount = 11`

dp[0] = 0
dp[1] = 1 (use coin 1)
dp[2] = 1 (use coin 2)
dp[3] = 2 (use coin `2 + coin` 1)
dp[4] = 2 (use coin `2 + coin` 2)
dp[5] = 1 (use coin 5)
dp[6] = 2 (use coin `5 + coin` 1)
...
dp[11] = 3 (use coin `5 + coin` `5 + coin` 1)
```

### TIME COMPLEXITY:
O(amount × len(coins))
For each amount from 1 to target, try all coins

### SPACE COMPLEXITY:
O(amount)
DP array of size amount + 1

### EDGE CASES:
- Amount is 0: return 0 (no coins needed)
- No solution possible: return -1
- Single coin type: straightforward calculation
- Large amounts: DP handles efficiently

</details>
"""

from collections import deque
from typing import Any


class Solution:
    def coinChange(self, coins: list[int], amount: int) -> int:
        """
        Approach: Dynamic Programming (Bottom-up)
        Time Complexity: O(amount * len(coins))
        Space Complexity: O(amount)
        """
        # Initialize DP array
        dp = [float("inf")] * (amount + 1)
        dp[0] = 0

        # For each amount from 1 to target
        for i in range(1, amount + 1):
            # Try each coin
            for coin in coins:
                if coin <= i:
                    dp[i] = min(dp[i], dp[i - coin] + 1)

        return int(dp[amount]) if dp[amount] != float("inf") else -1

    def coinChangeTopDown(self, coins: list[int], amount: int) -> int:
        """
        Approach: Dynamic Programming (Top-down with memoization)
        Time Complexity: O(amount * len(coins))
        Space Complexity: O(amount)
        """
        memo: dict[Any, Any] = {}

        def dp(n: Any) -> Any:
            # Base cases
            if n == 0:
                return 0
            if n < 0:
                return float("inf")

            # Check memo
            if n in memo:
                return memo[n]

            # Try each coin
            min_coins = float("inf")
            for coin in coins:
                result = dp(n - coin)
                if result != float("inf"):
                    min_coins = min(min_coins, result + 1)

            memo[n] = min_coins
            return min_coins

        result = dp(amount)
        return result if result != float("inf") else -1

    def coinChangeBFS(self, coins: list[int], amount: int) -> int:
        """
        Approach: BFS
        Time Complexity: O(amount * len(coins))
        Space Complexity: O(amount)
        """
        if amount == 0:
            return 0

        queue = deque([0])
        visited = {0}
        level = 0

        while queue:
            level += 1
            size = len(queue)

            for _ in range(size):
                current = queue.popleft()

                for coin in coins:
                    next_amount = current + coin

                    if next_amount == amount:
                        return level

                    if next_amount < amount and next_amount not in visited:
                        visited.add(next_amount)
                        queue.append(next_amount)

        return -1


# Test cases
if __name__ == "__main__":
    solution = Solution()

    # Test case 1
    coins1 = [1, 2, 5]
    amount1 = 11
    print(f"Input: coins = {coins1}, amount = {amount1}")
    print(f"Output: {solution.coinChange(coins1, amount1)}")  # 3

    # Test case 2
    coins2 = [2]
    amount2 = 3
    print(f"Input: coins = {coins2}, amount = {amount2}")
    print(f"Output: {solution.coinChange(coins2, amount2)}")  # -1

    # Test case 3
    coins3 = [1]
    amount3 = 0
    print(f"Input: coins = {coins3}, amount = {amount3}")
    print(f"Output: {solution.coinChange(coins3, amount3)}")  # 0

File: solutions/dynamic-programming/alternatives/322-coin-change.js
/**
 * 322. Coin Change
 * Medium
 *
 * Coin Change - Dynamic Programming Solution LeetCode 322: https://leetcode.com/problems/coin-change/ Problem: Given an array of coins and a target amount, return the fewest number of coins needed to make up that amount. Return -1 if the amount cannot be made up by any combination of the coins. @param {number[]} coins - Array of coin denominations @param {number} amount - Target amount to make @return {number} - Minimum number of coins needed, or -1 if impossible
 *
 * <details>
 * <summary><b>🔍 SOLUTION EXPLANATION</b></summary>
 *
 * ### INTUITION:
 * The key insight for solving Coin Change is to understand the core problem pattern
 * and apply the most efficient algorithmic approach.
 *
 * ### APPROACH:
 * 1. Analyze the problem requirements
 * 2. Choose the optimal data structure
 * 3. Implement the solution step by step
 * 4. Handle edge cases appropriately
 *
 * ### WHY THIS WORKS:
 * This approach works because it leverages the fundamental properties of the problem
 * to achieve an efficient solution.
 *
 * ### EXAMPLE WALKTHROUGH:
 * For a typical input, the algorithm processes the data systematically
 * to produce the expected output.
 *
 * </details>
 */

/**
 * Coin Change - Dynamic Programming Solution
 * LeetCode 322: https://leetcode.com/problems/coin-change/
 * 
 * Problem: Given an array of coins and a target amount, return the fewest number
 * of coins needed to make up that amount. Return -1 if the amount cannot be made
 * up by any combination of the coins.
 * 
 * @param {number[]} coins - Array of coin denominations
 * @param {number} amount - Target amount to make
 * @return {number} - Minimum number of coins needed, or -1 if impossible
 */

const coinChange = function(coins, amount) {
    // Edge cases
    if (amount === 0) return 0;
    if (!coins || coins.length === 0) return -1;
    
    // Create DP array initialized with amount + 1 (impossible value)
    // dp[i] represents the minimum coins needed to make amount i
    const dp = new Array(amount + 1).fill(amount + 1);
    
    // Base case: 0 amount needs 0 coins
    dp[0] = 0;
    
    // For each amount from 1 to target amount
    for (let currentAmount = 1; currentAmount <= amount; currentAmount++) {
        // Try each coin
        for (const coin of coins) {
            // If the coin value is less than or equal to current amount
            if (coin <= currentAmount) {
                // Take minimum of current solution and solution after using this coin
                dp[currentAmount] = Math.min(
                    dp[currentAmount],
                    dp[currentAmount - coin] + 1
                );
            }
        }
    }
    
    // If no solution found, dp[amount] will still be amount + 1
    return dp[amount] === amount + 1 ? -1 : dp[amount];
};

/**
 * Helper function to test the solution
 * @param {number[]} coins - Array of coin denominations
 * @param {number} amount - Target amount
 * @param {number} expected - Expected result
 */
function testCoinChange(coins, amount, expected) {
    const result = coinChange(coins, amount);
    console.log(`Input: coins = [${coins}], amount = ${amount}`);
    console.log(`Expected: ${expected}, Got: ${result}`);
    console.log(`Test ${result === expected ? 'PASSED' : 'FAILED'}\n`);
}

// Test cases
function runTests() {
    testCoinChange([1, 2, 5], 11, 3);           // Should return 3 (5 + 5 + 1)
    testCoinChange([2], 3, -1);                 // Should return -1 (impossible)
    testCoinChange([1], 0, 0);                  // Should return 0
    testCoinChange([1], 1, 1);                  // Should return 1
    testCoinChange([1, 2, 5, 10], 27, 4);       // Should return 4 (10 + 10 + 5 + 2)
}

// Run tests if this file is being run directly
if (require.main === module) {
    runTests();
}

// Export for use in other files
module.exports = {
    coinChange
};